src/rgw/rgw_dedup.cc

```diff
@@ -0,0 +1,2363 @@
+                                      md5_stats_t         *p_stats /* IN-OUT */)
+  {
+    bool should_print_debug = cct->_conf->subsys.should_gather<ceph_subsys_rgw, 20>();
+    if (unlikely(should_print_debug)) {
+      try_deduping_record_dbg(dpp, p_tgt_rec, block_id, rec_id, md5_shard);
+    }
+    //p_stats->processed_objects ++;
+    if (p_tgt_rec->s.flags.has_shared_manifest()) {
+      // record holds a shared_manifest object so can't be a dedup target
+      p_stats->skipped_shared_manifest++;
+      ldpp_dout(dpp, 20) << __func__ << "::skipped shared_manifest" << dendl;
+      return 0;
+    }
+    uint32_t size_4k_units = byte_size_to_disk_blocks(p_tgt_rec->s.obj_bytes_size);
+    //uint64_t on_disk_byte_size = disk_blocks_to_byte_size(size_4k_units);
+    key_t key(p_tgt_rec->s.md5_high, p_tgt_rec->s.md5_low, size_4k_units,
+              p_tgt_rec->s.num_parts);
+    dedup_table_t::value_t src_val;
+    int ret = p_table->get_val(&key, &src_val);
+    if (ret != 0) {
+      // record has no valid entry in table because it is a singleton
+      // should never happened since we purged all singletons before
+      ldpp_dout(dpp, 5) << __func__ << "::skipped singleton::" << p_tgt_rec->bucket_name
+                        << "/" << p_tgt_rec->obj_name << "::num_parts=" << p_tgt_rec->s.num_parts
+                        << "::ETAG=" << std::hex << p_tgt_rec->s.md5_high
+                        << p_tgt_rec->s.md5_low << std::dec << dendl;
+      ceph_abort("Unexpcted singleton");
+      return 0;
+    }
+
+    disk_block_id_t src_block_id = src_val.block_idx;
+    record_id_t src_rec_id = src_val.rec_id;
+    if (block_id == src_block_id && rec_id == src_rec_id) {
+      // the table entry point to this record which means it is a dedup source so nothing to do
+      p_stats->skipped_source_record++;
+      ldpp_dout(dpp, 20) << __func__ << "::skipped source-record" << dendl;
+      return 0;
+    }
+
+    // This records is a dedup target with source record on source_block_id
+    disk_record_t src_rec;
+    ret = load_record(d_dedup_cluster_ioctx, &src_rec, src_block_id, src_rec_id, md5_shard, &key, dpp);
+    if (unlikely(ret != 0)) {
+      p_stats->failed_src_load++;
+      // we can withstand most errors moving to the next object
+      ldpp_dout(dpp, 5) << __func__ << "::ERR: Failed load_record("
+                        << src_block_id << ", " << src_rec_id << ")" << dendl;
+      return 0;
+    }
+
+    ldpp_dout(dpp, 20) << __func__ << "::SRC=" << src_rec.bucket_name << "/" << src_rec.obj_name << dendl;
+    // verify that SRC and TGT records don't refer to the same physical object
+    // This could happen in theory if we read the same objects twice
+    if (src_rec.obj_name == p_tgt_rec->obj_name && src_rec.bucket_name == p_tgt_rec->bucket_name) {
+      p_stats->duplicate_records++;
+      ldpp_dout(dpp, 10) << __func__ << "::WARN: Duplicate records for object=" << src_rec.obj_name << dendl;
```

Comment:
Shouldn't the warning and error bits here and above be done by the logging subsystem, vs hardcoded in the strings?

Reply:
The logging subsystem has no concept of INFO/WARN/ERROR and instead is using numeric values from 0-255.
I need a way to quickly search the logs for warning/errors reported by my code